L1Trigger/DemonstratorTools/src/utilities.cc

```diff
@@ -0,0 +1,385 @@
+    while (getline(file, line)) {
+      if (line.empty())
+        continue;
+      if (line[0] == '#')
+        continue;
+
+      if (line.rfind("Board ", 0) != std::string::npos) {
+        id = line.substr(6);
+        break;
+      } else
+        throw std::logic_error("Found unexpected line found when searching for board ID: \"" + line + "\"");
+    }
+
+    // 2) Search for column labels (i.e. list of channels/links)
+    searchAndTokenize(file, "Quad/Chan :");
+    const auto tokens = searchAndTokenize(file, "Link :");
+    std::vector<size_t> channels;
+    std::transform(tokens.begin(), tokens.end(), std::back_inserter(channels), [](const std::string& s) {
+      return std::stoull(s);
+    });
+
+    // 3) Read the main data rows
+    const std::regex delimiterRegex("\\s+");
+    static const std::regex frameRegex("([01]s)?([01]v)([0-9a-fA-F]{16})");
+    std::vector<std::vector<Frame>> dataRows;
+    while (file.good() and getline(file, line)) {
+      if (line.empty() or line[0] == '#')
+        continue;
+
+      std::ostringstream prefixStream;
+      prefixStream << "Frame ";
+      prefixStream << std::setw(4) << std::setfill('0') << dataRows.size();
+      prefixStream << " :";
+
+      const std::string prefix(prefixStream.str());
+      if (line.rfind(prefix, 0) == std::string::npos)
+        throw std::logic_error("Found unexpected line found when searching for \"" + prefix + "\": \"" + line + "\"");
+
+      std::vector<l1t::demo::Frame> row;
+      std::sregex_token_iterator it(line.begin() + prefix.size(), line.end(), delimiterRegex, -1);
+      for (; it != std::sregex_token_iterator(); it++) {
+        const std::string token(it->str());
+        if (token.empty())
+          continue;
+
+        std::smatch what;
+        if (not std::regex_match(token, what, frameRegex))
+          throw std::logic_error("Token '" + token + "' doesn't match the valid format");
+
+        l1t::demo::Frame value;
+        // Import strobe if the strobe group is matched
+        if (what[1].matched) {
+          value.strobe = (what[1] == "1s");
+        }
+
+        value.valid = (what[2] == "1v");
+        value.data = ap_uint<64>(std::stoull(what[3].str(), nullptr, 16));
+
+        row.push_back(value);
+      }
+
+      dataRows.push_back(row);
+    }
+
+    return createBoardDataFromRows(id, channels, dataRows);
+  }
+
+  BoardData readX20File(std::istream& file, const FileFormat format) {
+    throw std::runtime_error("Reading X20 file format not yet implemented. Will be done ASAP.");
+  }
+
+  void writeAPxFile(const BoardData&, std::ostream&, const FileFormat);
+
+  void writeEMPFile(const BoardData&, std::ostream&, const FileFormat);
+
+  void writeX20File(const BoardData&, std::ostream&, const FileFormat);
+
+  void write(const BoardData& data, const std::string& filePath, const FileFormat format) {
+    // Open file
+    std::cout << "Writing board data (" << std::distance(data.begin(), data.end()) << " channels, "
```

Comment:
You should `edm::MessageLogger` rather than `std::cout`.

Reply:
When the code is built in CMSSW, the messages are now sent to `edm::MessageLogger`